Feature/iteration 1.2

22DPVEmer opened the pull request:
Implemented CRUD functionality for articles and user permissions can be granted, permission based access to resources is implemented. Images can also be handled in articles. Users can ask for permission to the admins, any single admin can accept or deny based on reason. Only the person who wrote an article can edit and delete article.
1. Image of article list:
![image](https://github.com/user-attachments/assets/c099d59a-c136-4bfd-b49b-e60486477ff9)
2. Image of create view for articles:
![image](https://github.com/user-attachments/assets/3ca57f68-cb92-4776-866a-805413e5d99b)
3. Single article view:
![image](https://github.com/user-attachments/assets/553390a8-7eaf-4213-9db2-55825bc727b1)
4. Delete confirmation view:
![image](https://github.com/user-attachments/assets/20593efb-0ee4-4cf2-be2b-c42bc13d71d0)
5. Edit article view:
![image](https://github.com/user-attachments/assets/97272bd2-e5f8-4472-bf6b-44d552cea96d)
6. Permission view for admin:
![image](https://github.com/user-attachments/assets/a714fc8e-c7cd-4dc4-857c-5bf77daadc34)
7. Permission view for basic user that is not confirmed yet to write articles:
![image](https://github.com/user-attachments/assets/4f81d9cf-cdcc-43c6-93be-b3f1fe32c765)

22DPVEmer committed:
feat: Add profile picture upload and user management improvements

- Implement Firebase storage integration for profile pictures
- Add client-side image preview and validation
- Enhance user profile management with profile picture support
- Update user entity and profile management page with new features
- Add Firebase and cloud storage dependencies
- Improve user profile editing experience

22DPVEmer committed:
chore: Update Firebase configuration and TinyMCE API key integration

22DPVEmer committed:
feat: Enhance database seeding with additional users and articles

- Add DbSet for Comments, Reports, Ranks, and PermissionRequests
- Seed 5 additional regular users without article writing permissions
- Add 10 sample articles with varied content and authors
- Configure PermissionRequest relationships with user navigation properties
- Update layout to include permissions navigation for authenticated users

22DPVEmer committed:
refactor: Modularize article creation and editing frontend code

- Extract CSS styles to separate files for Create, Edit, and Details views
- Create a centralized constants.js for shared configuration
- Implement a unified create-article.js for handling article form interactions
- Improve code organization and reduce duplication in article-related frontend scripts
- Add support for both create and edit modes in article form handling

22DPVEmer committed:
feat: Implement article management services and constants

- Add IArticleService and related implementations for article CRUD operations
- Introduce IPermissionService for handling user permission requests
- Create ArticleConstants and LogConstants for centralized message management

sreeharihm commented on Blog_App-iteration_1.1/Blog.Core/Services/ArticleService.cs:
```diff
@@ -0,0 +1,237 @@
+        private readonly ApplicationDbContext _context;
+        private readonly IFirebaseStorageService _firebaseStorageService;
+        private readonly ILogger<ArticleService> _logger;
+        private const int MAX_TOTAL_SIZE = 10 * 1024 * 1024; // 10MB in bytes
```
I would better add size in appsettings.config

22DPVEmer replied:
Moved it to appsettings.

22DPVEmer replied:
Moved it.

sreeharihm commented on Blog_App-iteration_1.1/Blog.Core/Services/ArticleService.cs:
```diff
@@ -0,0 +1,237 @@
+            var today = DateTime.UtcNow.Date;
+            switch (dateFilter)
+            {
+                case "today":
```
better move this filter to enum

22DPVEmer replied:
Moved it.

sreeharihm commented on Blog_App-iteration_1.1/Blog.Core/Services/ArticleService.cs:
```diff
@@ -0,0 +1,237 @@
+            if (model.FeaturedImageFile != null && model.FeaturedImageFile.Length > 0)
+            {
+                // Check file type
+                var allowedTypes = new[] { "image/jpeg", "image/png", "image/gif", "image/webp" };
```
all these type should be in the appsettings

22DPVEmer replied:
Now I use types from appsettings.

sreeharihm commented on Blog_App-iteration_1.1/Blog.Core/Services/ArticleService.cs:
```diff
@@ -0,0 +1,237 @@
+                }
+
+                // Check file size (limit to 5MB)
+                if (model.FeaturedImageFile.Length > 5 * 1024 * 1024)
```
some place it is 10 mb and here its 5 mb, any specific reason

22DPVEmer replied:
Changed it back to 10.

sreeharihm commented on Blog_App-iteration_1.1/Blog.Core/Services/ArticleService.cs:
```diff
@@ -0,0 +1,237 @@
+
+        public async Task<string> UploadImageAsync(IFormFile file, bool isFeatured)
+        {
+            if (file == null || file.Length == 0)
```
I would prefer fileupload should happend in background service

22DPVEmer replied:
Made a background service in services for that specifically now it does not interfere with loading speed.

sreeharihm commented on Blog_App-iteration_1.1/Blog.Web/Controllers/ArticlesController.cs:
```diff
@@ -0,0 +1,239 @@
+
+        // GET: Articles - Allow all users to view articles
+        [HttpGet]
+        [Route("")]
```
If not required, please remove it

22DPVEmer replied:
Removed it.

sreeharihm reviewed (commented):
Added review comments

22DPVEmer committed:
fix: implemented background service for images instead of direct processing and reorganized the code

22DPVEmer committed:
feat: Implemented asynchronous image processing with background service integration. Enhanced image handling with proper cleanup on article deletion. Fixed progress tracking for article image uploads and improved overall upload experience with proper size validation and progress feedback.

sreeharihm commented on Blog_App-iteration_1.1/Blog.Core/Services/ArticleService.cs:
```diff
@@ -0,0 +1,243 @@
+                Intro = model.Intro,
+                UserId = user.Id,
+                CreatedAt = DateTime.UtcNow,
+                IsPublished = true,
```
What is the use of this property IsPublished?

22DPVEmer replied:
It was an old idea, I will remove it as all articles are published, public.

sreeharihm commented on Blog_App-iteration_1.1/Blog.Core/Services/ArticleService.cs:
```diff
@@ -0,0 +1,243 @@
+            await _context.SaveChangesAsync();
+
+            // Handle featured image upload if provided
+            if (model.FeaturedImageFile != null && model.FeaturedImageFile.Length > 0)
```
Simplify it with string.isnullorempty or isnullorwhitespace

22DPVEmer replied:
the model.FeaturedImageFile is not a string so those methods will not work.

sreeharihm commented on Blog_App-iteration_1.1/Blog.Core/Services/ArticleService.cs:
```diff
@@ -0,0 +1,243 @@
+            // Handle featured image upload if provided
+            if (model.FeaturedImageFile != null && model.FeaturedImageFile.Length > 0)
+            {
+                ValidateImageFile(model.FeaturedImageFile);
```
Can we do this empty check in the validate method

22DPVEmer replied:
Moved the check it into the ValidateImageFile method.

sreeharihm commented on Blog_App-iteration_1.1/Blog.Core/Services/ArticleService.cs:
```diff
@@ -0,0 +1,243 @@
+
+                // Delete old featured image if it exists
+                if (!string.IsNullOrEmpty(existingArticle.FeaturedImage) && 
+                    existingArticle.FeaturedImage != "processing")
```
Please remove the hardcoding

22DPVEmer replied:
Revmoved all the hardcoding that I found.

sreeharihm commented on Blog_App-iteration_1.1/Blog.Core/Services/ArticleService.cs:
```diff
@@ -0,0 +1,243 @@
+                return false;
+            }
+
+            if (article.UserId != user.Id)
```
This should be come 1st before null check

22DPVEmer replied:
I combined it into a single if check with && instead.

sreeharihm reviewed (commented):
Added review comments

22DPVEmer committed:
feat: Enhance article service and constants for image handling

- Introduced new constants for image status and types in ArticleConstants
- Updated ArticleService to utilize new constants for image processing
- Improved error handling during image uploads and validations
- Removed IsPublished property from Article entity and related seed data

22DPVEmer committed:
refactor: Simplify image upload handling in ArticleService

- Streamlined error handling for featured image uploads by consolidating try-catch blocks
- Updated validation logic to skip checks for null files
- Improved code readability and maintainability in image processing logic
- Added more constants

sreeharihm reviewed (approved):
Looks good!
